### PR TITLE
Remove nops

### PR DESCRIPTION
**Pull request by sydgibs**
Done but needs a full system build to sanity check that I can't do locally. @yizhou7 can you make sure we can get a full build?

-- Adds analysis for when a NOP is necessary to prevent a LOOP_ERROR as per https://docs.opentitan.org/hw/ip/otbn/doc/#loop-nesting
-- Adds test framework and unit test suite to test analysis code
-- Adds required NOPs directly to implementation of modexp instead of relying on additions in the printer
-- Separates printing routine for modexp from the printer itself

### CODE REVIEW

File: arch/otbn/machine.s.dfy [hunk 1]
```diff
@@ -382,6 +383,99 @@ module ot_machine {
...
+
+predicate method has_overlap(c:simple_code) 
+{
+  match c
+    case SIns => false
+    case SJump => false
+    case SBranch(s) => false
+    case SWhile(s) =>
+      if |s| == 0 then false
+      else if (s[|s|-1].SWhile? || s[|s|-1].SBranch? || s[|s|-1].SJump?) then true // Should use Seq.last(s) here
```
Comment (parno) [inline arch/otbn/machine.s.dfy:421]: Should we go ahead and pull in the standard Sequence library for use here?

File: arch/otbn/machine.s.dfy [hunk 2]
```diff
@@ -382,6 +383,99 @@ module ot_machine {
...
+  if |s| == 0 then false
+  else has_overlap(s[0]) || has_overlap_seq(s[1..])
+}
+
+predicate method while_overlap(c:code) 
+{
+  has_overlap_seq(simplify_code(c))
+}
+
+/*
```
Comment (parno) [inline arch/otbn/machine.s.dfy:436]: Safe to remove all of the commented code at this point?

File: arch/otbn/modexp_printer.s.dfy [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import opened otbn_printer
+
+method Main()
+{
+    var p := new Printer({"modexp_var_3072_f4", "montmul"});
+
+    reveal va_code_modexp_var_3072_f4();
+    var c := va_code_modexp_var_3072_f4();
+
+    var n :=  while_overlap(c);
```
Comment (parno) [inline arch/otbn/modexp_printer.s.dfy:18]: This overlap check seems like it's probably redundant with the checks that happen inside `printTopLevelProc`, right?

File: arch/otbn/printer.s.dfy [hunk 4]
```diff
@@ -403,23 +404,16 @@ class Printer {
...
                 printIndent(); print("ret\n\n");
             }
             i := i + 1;
         }
     }
 }
-
-method Main()
-{
-    var comment := "/*\n  This code is generated by the veri-titan project: https://github.com/secure-foundations/veri-titan\n*/\n";
```
Comment (parno) [inline arch/otbn/printer.s.dfy:416]: This comment seems like it might be useful to retain.  Can we add it to `printTopLevelProc`?

**Review by parno — COMMENTED** [review]
Generally looks pretty good.  I added a few comments/questions.

File: arch/otbn/printer.s.dfy (hunk 4 above)
Comment (yizhou7) [reply arch/otbn/printer.s.dfy:416]: Sure, actually moved it into to `modexp_printer`.

**Pull request merged**